src/main/java/seedu/duke/userinterface/CliUserInterface.java

```diff
@@ -1,42 +1,58 @@
 package seedu.duke.userinterface;
 
-import seedu.duke.exceptions.InvalidCommandException;
-import seedu.duke.tasks.TaskList;
-import seedu.duke.userinterface.command.Add;
+import seedu.duke.storage.Storage;
 import seedu.duke.userinterface.command.CliCommand;
+import seedu.duke.userinterface.command.Exit;
 
-import java.lang.String;
 import java.util.Scanner;
 
 public class CliUserInterface {
-    private static int numberOfTasks;
-    private static TaskList list;
-    private static AppState appState;
-    public Add add;
-
-    public CliUserInterface(TaskList list, int numberOfTasks) {
-        this.list = list;
-        this.numberOfTasks = numberOfTasks;
+    private AppState appState;
+
+    private boolean toQuit = false;
+
+    public CliUserInterface() {
+
     }
 
-    public void executeCommand(String userInput) throws Exception {
+    private void loadState() {
+        Storage storage = new Storage();
+        appState = storage.readFromFile();
+    }
 
+    private void saveState() {
+        Storage storage = new Storage();
+        storage.saveToFile(appState);
+    }
+
+    public void run() {
+        startUI();
+        loadState();
+        String userInput;
+        Scanner keyboard = new Scanner(System.in);
+        while (!toQuit) {
+            userInput = keyboard.nextLine();
+            if (userInput.equals(Exit.COMMAND_WORD)) {
+                saveState();
+                toQuit = true;
+            }
+            try {
+                executeCommand(userInput);
+            } catch (Exception e) {
+                // TODO: actually make exceptions that make sense
+                e.printStackTrace();
```

Comment:
InvalidInputException?

Reply:
I decided to not worry about exceptions too much at this stage. There is a better way to do exceptions that I want to discuss later. For now this should be ok as it causes the program to not crash.

Reply:
Okay, sure!